network/ssh/reachable.go

```diff
@@ -0,0 +1,216 @@
+	FindHost(hostPorts []network.HostPort, publicKeys []string) (network.HostPort, error)
+}
+
+// NewReachableChecker creates a ReachableChecker that can be used to check for
+// Hosts that are viable SSH targets.
+// When FindHost is called, we will dial the entries in the given hostPorts, in
+// parallel, using the given dialer, closing successfully established
+// connections after checking the ssh key. Individual connection errors are
+// discarded, and an error is returned only if none of the hostPorts can be
+// reached when the given timeout expires.
+// If publicKeys is a non empty list, then the SSH host public key will be
+// checked. If it is not in the list, that host is not considered valid.
+//
+// Usually, a net.Dialer initialized with a non-empty Timeout field is passed
+// for dialer.
+func NewReachableChecker(dialer Dialer, timeout time.Duration) *reachableChecker {
+	return &reachableChecker{
+		dialer:  dialer,
+		timeout: timeout,
+	}
+}
+
+// hostKeyChecker checks if this host matches one of allowed public keys
+// it uses the golang/x/crypto/ssh/HostKeyCallback to find the host keys on a
+// given connection.
+type hostKeyChecker struct {
+	// AcceptedKeys is a set of the Marshalled PublicKey content.
+	AcceptedKeys set.Strings
+	// Stop will be polled for whether we should stop trying to do any work
+	Stop <-chan struct{}
+	// HostPort is the identifier that corresponds to this connection
+	HostPort network.HostPort
+	// Accepted will be passed HostPort if it validated the connection
+	Accepted chan network.HostPort
+	// Dialer is a Dialer that allows us to initiate the underlying TCP connection
+	Dialer Dialer
+	// Finished will be set an event when we've finished our check (success or failure)
+	Finished chan struct{}
+}
+
+var hostKeyNotInList = errors.New("host key not in expected set")
+var hostKeyAccepted = errors.New("host key was accepted, retry")
+
+func (h *hostKeyChecker) hostKeyCallback(hostname string, remote net.Addr, key ssh.PublicKey) error {
+	// Note: we don't do any advanced checking of the PublicKey, like whether
+	// the key is revoked or expired. All we care about is whether it matches
+	// the public keys that we consider acceptable
+	authKeyForm := ssh.MarshalAuthorizedKey(key)
+	debugName := hostname
+	if hostname != remote.String() {
+		debugName = fmt.Sprintf("%s at %s", hostname, remote.String())
+	}
+	logger.Tracef("checking host key for %s, with key %q", debugName, authKeyForm)
+
+	lookupKey := string(key.Marshal())
+	if len(h.AcceptedKeys) == 0 || h.AcceptedKeys.Contains(lookupKey) {
+		logger.Debugf("accepted host key for: %s", debugName)
+		// This key was valid, so return it, but if someone else was found
+		// first, still exit.
+		select {
+		case h.Accepted <- h.HostPort:
+		case <-h.Stop:
+		}
+		return hostKeyAccepted
+	}
+	logger.Debugf("host key for %s not in our accepted set: use --debug --log-level=TRACE to see actual key", debugName)
```

Comment:
logger.Debugf("host key for %s not in our accepted set: log at TRACE to see raw keys", debugName)

better?